packages/flutter_tools/test/commands.shard/permeable/upgrade_test.dart

```diff
@@ -113,15 +116,15 @@ void main() {
       );
       expect(await result, FlutterCommandResult.success());
       expect(testLogger.statusText, contains('A new version of Flutter is available'));
-      expect(testLogger.statusText, contains(fakeCommandRunner.remoteRevision));
-      expect(testLogger.statusText, contains(revision));
+      expect(testLogger.statusText, contains('The latest version: null (revision def456)'));
```

Comment:
Can you change this from `null` to an actual version tag? Your update to FakeUpgradeCommandRunner.fetchLatestVersion is constructing a new `MockFlutterVersion` on the fly, which is returning `null` when `.frameworkVersion` is invoked.

Reply:
> Probably the easiest way to fix this is to set a new property on `FakeUpgradeCommandRunner` called `FlutterVersion frameworkVersion;` that can be optionally set in the constructor.

Could you elaborate how this would solve the problem? My understanding is that `MockFlutterVersion.frameworkVersion` is returning null as it is not assigned, and can be fixed by the `when`-`thenReturn` construct just like `frameworkRevision`.

Reply:
In the tool code, we explicitly set mock return values in the test code, so that the mocked behavior is obvious when reading the test. It will make updating the test easier if, for example, the behavior of `fetchLatestVersion()` were to change in a future test.

Something like this:

```
class FakeUpgradeCommandRunner extends UpgradeCommandRunner {
  bool willHaveUncommittedChanges = false;
  bool alreadyUpToDate = false;

  String remoteRevision;
  FlutterVersion remoteVersion;

  @override
  Future<FlutterVersion> fetchLatestVersion() async => remoteVersion;
}
```

Reply:
> In the tool code, we explicitly set mock return values in the test code

This snippet alongwith snippets in my previous comment show how that can be done:
```dart
const String revision = 'abc123';
const String upstreamRevision = 'def456';
const String version = '1.2.3';
const String upstreamVersion = '4.5.6';
when(flutterVersion.frameworkRevision).thenReturn(revision);
when(flutterVersion.frameworkRevisionShort).thenReturn(revision);
fakeCommandRunner.remoteVersion = upstreamVersion;
fakeCommandRunner.remoteRevision = upstreamRevision;
```
Also I have some issues with the snippet you've provided:
```dart
String remoteRevision;
FlutterVersion remoteVersion;

@override
Future<FlutterVersion> fetchLatestVersion() async => remoteVersion;
```
 - Making this change would require all tests to explicitly define `FakeUpgradeCommandRunner.remoteVersion`.
 - What do we assign to `FakeUpgradeCommandRunner.remoteVersion` in the tests, `MockFlutterVersion` or `FlutterVersion`?
 - If we assign `MockFlutterVersion`, we can just simply, for example, do:
```dart
MockFlutterVersion latestVersion = MockFlutterVersion();
when(latestVersion.frameworkRevision).thenReturn(upstreamRevision);
when(latestVersion.frameworkRevisionShort).thenReturn(upstreamRevision);
when(latestVersion.frameworkVersion).thenReturn(upstreamVersion);
fakeCommandRunner.remoteVersion = latestVersion;
```
or
```dart
fakeCommandRunner.remoteVersion = MockFlutterVersion(revision: remoteRevision, version: remoteVersion);
```
Having `FakeUpgradeCommandRunner.remoteRevision` will be redundant in either case.
- If we assign `FlutterVersion`, `processManager` would have additional instructions, and the expectation will fail. If we assign `MockFlutterVersion`, it will not have a "version" tagged by default unless we explicitly provide one using either of the two above.

I'm sorry that I cannot follow your approach, but could you provide me an example on how `FakeUpgradeCommandRunner.remoteRevision` would be used in the tests?

Reply:
> > In the tool code, we explicitly set mock return values in the test code
> 
> This snippet alongwith snippets in my previous comment show how that can be done:
> 
> ```dart
> const String revision = 'abc123';
> const String upstreamRevision = 'def456';
> const String version = '1.2.3';
> const String upstreamVersion = '4.5.6';
> when(flutterVersion.frameworkRevision).thenReturn(revision);
> when(flutterVersion.frameworkRevisionShort).thenReturn(revision);
> fakeCommandRunner.remoteVersion = upstreamVersion;
> fakeCommandRunner.remoteRevision = upstreamRevision;
> ```
> 
> Also I have some issues with the snippet you've provided:
> 
> ```dart
> String remoteRevision;
> FlutterVersion remoteVersion;
> 
> @override
> Future<FlutterVersion> fetchLatestVersion() async => remoteVersion;
> ```
> 
> * Making this change would require all tests to explicitly define `FakeUpgradeCommandRunner.remoteVersion`.

If `runner.remoteVersion` isn't read in the test, it does not need to be set, right? We can leave it null. If it IS read during the test, we should force the test author to explicitly set the property with a mock object, to ensure that we don't get false positives for new tests. Where possibly, we try to write our tests so that behavior that isn't explicitly set will fail the test, so that the author gets signal that they need to mock a certain behavior.

> * What do we assign to `FakeUpgradeCommandRunner.remoteVersion` in the tests, `MockFlutterVersion` or `FlutterVersion`?
> * If we assign `MockFlutterVersion`, we can just simply, for example, do:
> 
> ```dart
> MockFlutterVersion latestVersion = MockFlutterVersion();
> when(latestVersion.frameworkRevision).thenReturn(upstreamRevision);
> when(latestVersion.frameworkRevisionShort).thenReturn(upstreamRevision);
> when(latestVersion.frameworkVersion).thenReturn(upstreamVersion);
> fakeCommandRunner.remoteVersion = latestVersion;
> ```
> 
> or
> 
> ```dart
> fakeCommandRunner.remoteVersion = MockFlutterVersion(revision: remoteRevision, version: remoteVersion);
> ```

Yes, the former is more verbose. We favor longer tests with explicit mocked behavior. The former example is easier for a contributor to read and know exactly what is mocked test behavior. Consider, for example, a future test wants to ensure that a `frameworkRevisionShort` is a subset of `frameworkRevision`. It will be clear in the former implementation that both values are manually mocked, and how to update the test.

> * If we assign `FlutterVersion`, `processManager` would have additional instructions, and the expectation will fail. If we assign `MockFlutterVersion`, it will not have a "version" tagged by default unless we explicitly provide one using either of the two above.

This is a good point. You can go either way, but passing a `MockFlutterVersion` and mocking the return value of `frameworkVersion` would be easier. This would still ensure that we're reading the version label and interpolating it into stdout. If you want to pass a real `FlutterVersion`, you would need to update `FakeProcessManager`, with the method `processManager.addCommands()`.